### DISCUSSION TITLE
Data_map - Javascript - Json parse

### DISCUSSION THREAD

**Discussion by Rbiaggio** [frontendbr/forum#1849] [Perguntas e Respostas]
boa noite. Eu preciso de uma grande ajuda.
Eu faço uma chamada à uma API e estava utilizando o código abaixo, porém, hoje parou de funcionar e não estou entendendo o motivo, visto que o json parece ser o mesmo. 

Estou usando o código abaixo:

    $(document).ready(function(){
      $("#button-buscar-licencas").click(function(){
        var urlapi = 'HTTP://API';
        var url=urlapi;
        var xhttp = new XMLHttpRequest();
        xhttp.open("GET", url, false);
        xhttp.send();
        var obj = JSON.parse(xhttp.responseText);
        var data_map = new Map();
        var index = 0;
          for (i = 0; i < obj.length; i++) {
          if(data_map.has(obj[i].hostGroup.name)) {
            data_map.set(obj[i].hostGroup.name, data_map.get(obj[i].hostGroup.name) + +obj[i].consumedHostUnits);
            } else {
                data_map.set(obj[i].hostGroup.name, +obj[i].consumedHostUnits);
              }
        }
        const objConverted = Object.fromEntries(data_map);
        var myJSON = JSON.stringify(objConverted);  
        data_map.forEach(mostraDados);   

        }

    function montaTabelaLicencas(hostgroup, licenca) {

      var newRow = $('<tr class="linha-table">');
          var cols = "";

            cols += '<td>' + licenca + '</td>';
            cols += '<td>' + hostgroup + '</td>';
            cols += '<td class="actions">';

          newRow.append(cols);

           $("#table-licencas").append(newRow); 
           return false;           
    };

    <div class="div-table-licencas" id="div-table-licencas">
      <button type="button" class="btn btn-search-licencas" id="button-buscar-licencas" title="Buscar por Ambiente"><i class="fa fa-search fa-1x" aria-hidden="true"></i><br></button><input class="form-control mb-4" id="tableSearch" type="text" placeholder="Filtrar">
      <table class="table table-sm table-hover table-bordered" id="table-licencas">
        <thead>
          <tr>
            <th>Sistema</th>
            <th>Licenças Utilizadas</th>
            <th>Ações</th>
          </tr>
        </thead>
        <tbody id="myTable">
        </tbody>
      </table>
    </div>

O problema é que não estou mais conseguindo pegar o hostGroup.name. Não sei o motivo. 

O retorno da API está abaixo:
[![inserir a descrição da imagem aqui][1]][1]

O erro é o abaixo:

[![inserir a descrição da imagem aqui][2]][2]

  [1]: https://i.stack.imgur.com/naJeS.png
  [2]: https://i.stack.imgur.com/bseHQ.png

Alguém consegue ajudar?

Obrigado.

**Comment by gabriel-hahn** [comment]
Buenas @Rbiaggio, aparentemente algum dos registros da tua lista que contém a propriedade `hostGroup` está vindo sem essa propriedade, gerando esse erro quando o `for` passa por esse cara e tenta acessar a propriedade `name`.
Como tu comentou que o código funcionava antes e não existiu nenhuma alteração de API ou da estrutura da resposta, bem provável que é somente esse detalhe. Se realmente for uma propriedade não obrigatória, tu pode simplesmente ajustar o teu `if` ou validar ela antes, como por exemplo:

```
const hostGroup = obj[i].hostGroup;
if (hostGroup && hostGroup.name) { ... }
```

Talvez só estou repetindo algo que tu já fez, mas coloca um `console.log` no `obj` e tenta verificar se essa propriedade realmente existe em todos os registros.

Espero ter ajudado :)

**Comment by Rbiaggio** [comment]
@gabriel-hahn Porra meu, acho que é isso mesmo. Alguns registros estão sem essa propriedade hostGroup. Alguém instalou sem passar esse parâmetro. Ele é obrigatório. 

Eu desabilitei esses caras, vamos ver se volta a funcionar. 

Existe alguma forma de fazer com que retorno somente os registro que TEM o hostGroup.name? 

Muito obrigado pela ajuda.

**Comment by gabriel-hahn** [comment]
Você pode solicitar uma alteração na API para fazer esse campo ser obrigatório ao salvar esses registros, retornar somente registros que contém essa propriedade ou fazer uma validação na lista de registros obtida do request no front-end mesmo, usando um `.filter`, por exemplo:

`const listWithHostGroup = obj.filter(item => !!item.hostGroup);`

Mas ai depende das tuas regras de negócio envolvidas. Se esse campo realmente é obrigatório, uma alteração na API para não permitir novos registros com o `hostGroup` vazio parece fazer mais sentido.

**Comment by Rbiaggio** [comment]
Então, o problema é que é o seguinte, esse campo é retornado pela API a partir de um agente instalado no S.O do servidor. Se quando a pessoa instalar o agente,  não passar esse parâmetro, ele não irá retornar o hostGroup.name. Foi uma ação indevida, um erro, não deveria ter acontecido. Eu desabilitei esses agentes, mas gostaria de colocar uma solução de contorno. 

Eu estou tentando fazer algo para ele percorrer tudo e pegar só o que tiver o hostGroup.name, mas não está rolando.

Vou tentar fazer esse filtro conforme sugeri.
